### PR TITLE
[WFLY-15075] Add encryption support to FilesystemSecurityRealm

### PR DESCRIPTION
**Pull request by Ashpan**
Jira: https://issues.redhat.com/browse/WFLY-15075

Requires: https://issues.redhat.com/browse/WFCORE-5530
WildFly Core PR with tests: https://github.com/wildfly/wildfly-core/pull/4983

### CODE REVIEW

**Comment by fjuma** [comment]
@Ashpan When you get a chance, these commits can be squashed.

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 1]
```diff
@@ -127,6 +127,28 @@ _jboss.server.config.dir_.
 If your directory is located outside of _jboss.server.config.dir_, then
 you need to change the _path_ and _relative-to_ values appropriately.
 
+It is possible to encrypt the filesystem realm by configuring the credential-store and secret-key attributes for the realm.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:130]: s/credential-store/`credential-store`
s/secret-key/`secret-key`

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 2]
```diff
@@ -127,6 +127,28 @@ _jboss.server.config.dir_.
...
+
+When creating a an encrypted filesystem realm, you will have to specify two more attributes.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:132]: s/a an/an

Comment (fjuma) [reply docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:132]: Would be good to reword this since only the `credential-store` is a required attribute. Perhaps something like:

To enable encryption for a `filesystem-realm`, there are a couple attributes that can be configured.

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 3]
```diff
@@ -127,6 +127,28 @@ _jboss.server.config.dir_.
...
+
+* `credential-store`: This attribute specifies the credential-store where the secret key used for encrypting and decrypting the realm is stored.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:134]: `credential-store`

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 4]
```diff
@@ -127,6 +127,28 @@ _jboss.server.config.dir_.
...
+
+* `secret-key`: This attribute specifies the alias to the secret key stored in the credential store. By default the alias in the credential store is set to `key`
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:136]: s/alias to the secret key.../alias within the `credential-store` that contains the secret key to be used to encrypt and decrypt the realm

Comment (fjuma) [reply docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:136]: s/By default the alias in the credential store is set to `key`/The default value is `key`.

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 5]
```diff
@@ -127,6 +127,28 @@ _jboss.server.config.dir_.
...
+
+To create a the credential store and secret key for the attributes above, you can create a secret-key-credential-store using the folloing commands. This will create a credential store called `mycredstore` with the secret key alias `key`.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:138]: Maybe this could be reworded to:

As an example, to create a `credential-store` that is automatically populated with a `SecretKey` under the alias `key`, the following command can be used:

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 6]
```diff
@@ -127,6 +127,28 @@ _jboss.server.config.dir_.
...
+[source,options="nowrap"]
+----
+/subsystem=elytron/secret-key-credential-store=mycredstore:add(path=propcredstore.cs, relative-to=jboss.server.config.dir, create=true, populate=true)
+----
+
+
+To create an encrypted filesystem realm with a credential-store called ``mycredstore`` and the secret-key alias ``key``, the following command would be used
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:145]: An encrypted `filesystem-realm` that makes use of the `credential-store` could then be created as follows:

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 7]
```diff
@@ -13,17 +13,19 @@ Every identity is stored in an XML file where the name of the file is the name o
 * `encoded` When encoding is set to true, the identity names will be BASE32 encoded before they are used as filenames. This is beneficial, because some filesystems are case-insensitive or might restrict set of characters allowed in a filename. Default value is true.
 * `hash-encoding` specifies the string format for the password if it is not stored in plain text. Set to BASE64 bt default, but HEX is also supported.
 * `hash-charset` specifies the character set to use when converting the password string to a byte array. Set to UTF-8 by default.
+* `credential-store` specifies the credential store where the secret key is stored if the filesystem realm is to be encrypted. This attribute is optional.
+* `secret-key` specifies the alias of the secret key that is stored in the previously specified credential store. This attribute is optional.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:17]: Would be good to mention the default value.

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 8]
```diff
@@ -13,17 +13,19 @@ Every identity is stored in an XML file where the name of the file is the name o
...
 
 Filesystem security realm can be added with WildFly management CLI or with Elytron API.
 
 === WildFly management CLI
 
 The following command creates realm with name `fsRealm` that resides in directory `fs-realm-users` inside the base configuration directory defined by `jboss.server.config.dir` property. The levels number is 2 and encoded value is true.
-The charset is UTF-8 and the string encoding is BASE64.
+The charset is UTF-8 and the string encoding is BASE64. The credential-store being referenced is called `mycredstore` and the alias of the key in that credential-store is `key`. The `key` secret key will be used to encrypt and decrypt the filesystem realm
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:24]: Does this belong in this section?

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 9]
```diff
@@ -13,17 +13,19 @@ Every identity is stored in an XML file where the name of the file is the name o
...
 
 [source,options="nowrap"]
 ----
-/subsystem=elytron/filesystem-realm=fsRealm:add(path=fs-realm-users, relative-to=jboss.server.config.dir)
+/subsystem=elytron/filesystem-realm=fsRealm:add(path=fs-realm-users, relative-to=jboss.server.config.dir, credential-store=mycredstore, secret-key=key)
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:28]: Same here.

**Review by fjuma — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc (hunk 1 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc]: Fixed

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc (hunk 3 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc]: Fixed

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc (hunk 5 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc]: Reworded

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc (hunk 6 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc]: Fixed

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc (hunk 7 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc]: Added

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc (hunk 8 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc]: Moved to a new line

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 10]
```diff
@@ -127,6 +127,27 @@ _jboss.server.config.dir_.
...
 
+It is possible to encrypt the filesystem realm by configuring the `credential-store` and `secret-key` attributes for the realm.
+
+To enable encryption for a `filesystem-realm`, there are a couple of attributes that can be configured.
+
+* `credential-store`: This attribute specifies the `credential-store` where the secret key used for encrypting and decrypting the realm is stored.
+
+* `secret-key`: This attribute specifies the alias within the `credential-store` that contains the secret key to be used to encrypt and decrypt the realm. The default value is `key`
+
+As an example, to create a `credential-store` that is automatically populated with a `SecretKey` under the alias key, the following command can be used:
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:138]: `key`

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 11]
```diff
@@ -127,6 +127,27 @@ _jboss.server.config.dir_.
 If your directory is located outside of _jboss.server.config.dir_, then
 you need to change the _path_ and _relative-to_ values appropriately.
 
+It is possible to encrypt the filesystem realm by configuring the `credential-store` and `secret-key` attributes for the realm.
+
+To enable encryption for a `filesystem-realm`, there are a couple of attributes that can be configured.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:132]: Just realized that this sentence rephrases the above sentence so maybe the above one can be removed.

Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:132]: Might be good to add a heading called Encryption or something like that to make this content stand out.

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 12]
```diff
@@ -13,6 +13,8 @@ Every identity is stored in an XML file where the name of the file is the name o
 * `encoded` When encoding is set to true, the identity names will be BASE32 encoded before they are used as filenames. This is beneficial, because some filesystems are case-insensitive or might restrict set of characters allowed in a filename. Default value is true.
 * `hash-encoding` specifies the string format for the password if it is not stored in plain text. Set to BASE64 bt default, but HEX is also supported.
 * `hash-charset` specifies the character set to use when converting the password string to a byte array. Set to UTF-8 by default.
+* `credential-store` specifies the credential store where the secret key is stored if the filesystem realm is to be encrypted. This attribute is optional.
+* `secret-key` specifies the alias of the secret key that is stored in the previously specified credential store. This attribute is optional. Set to `key` by default.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:17]: The `filesystem-realm` resource doesn't set a default value for `secret-key` so this last part can be removed.

File: docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc [hunk 13]
```diff
@@ -127,6 +127,27 @@ _jboss.server.config.dir_.
...
+
+* `credential-store`: This attribute specifies the `credential-store` where the secret key used for encrypting and decrypting the realm is stored.
+
+* `secret-key`: This attribute specifies the alias within the `credential-store` that contains the secret key to be used to encrypt and decrypt the realm. The default value is `key`
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/Using_the_Elytron_Subsystem.adoc:136]: The default portion here should be removed (see my comment below).

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 14]
```diff
@@ -47,11 +51,19 @@ To delete existing identity from filesystem realm:
 
 When creating filesystem realm with Elytron API, you can specify `name rewriter` in the constructor. The name rewriter will be applied to identity names to transform them from one form to another. It can be used to normalize the case, trim extra whitespaces, map one naming scheme to another, remove realm component from identity name (e.g. "user@realm" to "user") or to perform validation step on the syntax of the name.
 
-Class `org.wildfly.security.auth.realm.FileSystemSecurityRealm` is used to instantiate the realm.
+To encrypt the filesystem realm programmatically, you can specify a SecretKey when instantiating the realm.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:54]: s/To encrypt the filesystem realm/To enable encryption for a filesystem realm

Comment (fjuma) [reply docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:54]: `SecretKey`

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 15]
```diff
@@ -47,11 +51,19 @@ To delete existing identity from filesystem realm:
...
+
+Class `org.wildfly.security.auth.realm.FileSystemSecurityRealm` is used to instantiate the realm. A builder FileSystemSecurityRealm builder is available at `org.wildfly.security.auth.realm.FileSystemSecurityRealmBuilder` to make the instantiating process easier.
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:56]: Maybe we could reword this to something like:

A `org.wildfly.security.auth.realm.FileSystemSecurityRealm` can be instantiated using a `builder()` method as shown in the example below.

File: testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+                cli.sendLine(String.format("/subsystem=elytron/filesystem-realm=%s:add(path=%s,relative-to=jboss.server.config.dir, credential-store=%s, secret-key=%s)",
+                        realmName, path, credentialStore, secretKey));
+                cli.sendLine(String.format("/subsystem=elytron/filesystem-realm=%s:add-identity(identity=%s)", realmName, username));
+                cli.sendLine(String.format("/subsystem=elytron/filesystem-realm=%1$s:set-password(identity=%2$s, digest={algorithm=digest-md5, realm=%1$s, password=%3$s})",
+                        realmName, username, password));
+                cli.sendLine(String.format("/subsystem=elytron/filesystem-realm=%s:add-identity-attribute(identity=%s, name=Roles, value=[JBossAdmin])",
+                        realmName, username));
+                cli.sendLine(String.format("/subsystem=elytron/security-domain=%1$s:add(realms=[{realm=%2$s,role-decoder=from-roles-attribute}],default-realm=%2$s,permission-mapper=default-permission-mapper)",
+                        domainName, realmName));
+                cli.sendLine(String.format(
```
Comment (fjuma) [inline testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:146]: I think you can remove this line since it's not needed.

File: testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+                        "/subsystem=elytron/http-authentication-factory=%1$s:add(http-server-mechanism-factory=%2$s,security-domain=%1$s,"
+                                + "mechanism-configurations=[{mechanism-name=BASIC,mechanism-realm-configurations=[{realm-name=\"%1$s\"}]}])",
+                        domainName, PREDEFINED_HTTP_SERVER_MECHANISM_FACTORY));
+                cli.sendLine(String.format(
+                        "/subsystem=undertow/application-security-domain=%s:add(http-authentication-factory=%s)",
```
Comment (fjuma) [inline testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:151]: Here instead of `http-authentication-factory=...`, you can just use `security-domain=...`.

File: testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+                cli.sendLine(String.format("/subsystem=elytron/filesystem-realm=%s:write-attribute(name=credential-store, value=%s)",
+                        realmName, credentialStore));
+            }
+            ServerReload.reloadIfRequired(managementClient);
+        }
+
+        private void tearDownDomain(String deployment, String domainName, String realmName, String username, String credentialStore) throws Exception {
+            try (CLIWrapper cli = new CLIWrapper(true)) {
+                cli.sendLine(String.format("/subsystem=undertow/application-security-domain=%s:remove()", deployment));
+                cli.sendLine(String.format("/subsystem=elytron/http-authentication-factory=%s:remove()",
```
Comment (fjuma) [inline testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:190]: This line can be removed.

File: testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+
+    /**
+     *
+     * Test Filesystem realm when choosing to encrypt it with a bad Secret Key.
+     */
+    @Test
+    @OperateOnDeployment(DEPLOYMENT_ENCRYPTED)
+    public void testEncryptionSecretKeyFail(@ArquillianResource URL webAppURL) throws Exception {
+        URL url = prepareURL(webAppURL);
+        SetUpTask.setUpBadCredentialStore(SetUpTask.REALM_NAME, SetUpTask.CREDENTIAL_STORE+"New");
```
Comment (fjuma) [inline testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: Because the bad credential store has a different name from the good one, you should be able to do the setup in the regular `setup` method below. That way, you won't need a boolean to track whether or not it's been created.

**Review by fjuma — APPROVED** [review]

**Comment by OndrejKotek** [comment]
I miss a documentation for the conversion functionality of `elytron-tool` here: `filesystem-realm-encrypt`.

File: testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
```
Comment (OndrejKotek) [inline testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: I don't understand why this test passes sometimes - there is no `mycredstoreNew` resource, hence this test should fail here because the call (writing the attribute) should fail.

Comment (Ashpan) [reply testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: I changed the implementation of creating the secondary Credential Store resource to be in the `setUpTestDomain(...)` method. This way the resource is always present

https://github.com/ashpan/wildfly/blob/WFLY-15075-encrypt/testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java#L148

Comment (OndrejKotek) [reply testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: I'm sorry, I see the resource is set up at the last line of `setUpTestDomain`. That means that the whole test case fails sometimes, going deeper...

Comment (OndrejKotek) [reply testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: I suspect [LdapRealmTestCase](https://github.com/wildfly/wildfly/blob/3eaeac0b07b4a03917f94bfb8368b6ee09c12530/testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/LdapRealmTestCase.java#L340) of setting the `from-roles-attribute` without proper cleaning. I don't have time to verify this today, I can continue tomorrow if no one will take this over.

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 21]
```diff
@@ -21,9 +23,11 @@ Filesystem security realm can be added with WildFly management CLI or with Elytr
 The following command creates realm with name `fsRealm` that resides in directory `fs-realm-users` inside the base configuration directory defined by `jboss.server.config.dir` property. The levels number is 2 and encoded value is true.
 The charset is UTF-8 and the string encoding is BASE64.
 
+The credential-store being referenced is called `mycredstore` and the alias of the `secret-key` in that credential-store is `key`. The secret key will be used to encrypt and decrypt the filesystem realm
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:26]: `credential-store` instead of credential-store (2 places on this line)

Should add a period at the end of realm.

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 22]
```diff
@@ -232,3 +244,14 @@ $JBOSS_HOME/bin/elytron-tool.sh filesystem-realm -u conf/users.properties -r con
 ----
 
 This command creates new filesystem realm with users taken from users.properties file and roles taken from roles.properties file. Script `example-fs-realm.sh` that contains the commands for WildFly CLI is generated as well. The script adds this filesystem realm to the Elytron subsystem and also adds new security domain that uses this filesystem realm as a default realm.
+
+== Converting plain Filesystem realn into Encrypted Filesystem realm
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:248]: Converting an unencrypted filesystem realm into an encrypted filesystem realm

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 23]
```diff
@@ -232,3 +244,14 @@ $JBOSS_HOME/bin/elytron-tool.sh filesystem-realm -u conf/users.properties -r con
...
+
+Since the addition of Encrypted Filesystem Realm, Elytron has added a feature to convert plain unencrypted filesystem realms, to encrypted filesystem realm. You can use `Elytron Tool` to convert these realms. Below is an example of how to run this tool from the command line:
```
Comment (fjuma) [inline docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc:250]: I think we could simplify this to something like:

It is possible to convert an unencrypted filesystem realm into an encrypted one using the Elytron Tool. In particular, the `filesystem-realm-encrypt` command can be used as shown below:

File: testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java (hunk 20 above)
Comment (Ashpan) [reply testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: Removed `from-roles-attribute` resource during teardown of `LdapRealmTestCase`

https://github.com/Ashpan/wildfly/blob/6c25eb8dc9861961713ad7ce934b9864f8e4f1b2/testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/LdapRealmTestCase.java#L363

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc (hunk 21 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc]: Changed

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc (hunk 22 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc]: Changed

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc (hunk 23 above)
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc]: Changed

File: docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc [hunk 24]
```diff
@@ -232,3 +244,14 @@ $JBOSS_HOME/bin/elytron-tool.sh filesystem-realm -u conf/users.properties -r con
...
+
+[source,options="nowrap"]
+----
+$JBOSS_HOME/bin/elytron-tool.sh filesystem-realm-encrypt -i ./standalone/configuration/fs-realm-plain -o ./standalone/configuration/fs-realm-enc -c ./mycredstore.cs
+----
+
+This command creates new filesystem realm all data taken from the realm at `./standalone/configuration/fs-realm-plain` and stores the new realm at `./standalone/configuration/fs-realm-enc`, using a Secret Key from the credential store located at `./mycredstore.cs`. A `.cli` script will also be generated at the root of the filesystem realm that contains the commands for WildFly CLI. The script creates adds the `secret-key-credential-store` resource and this filesystem realm to the Elytron subsystem.
```
[the remark this answers is not in the document]
Comment (Ashpan) [reply docs/src/main/asciidoc/_elytron/components/Filesystem_Security_Realm.adoc]: Changed

**Review by fjuma — APPROVED** [review]

File: testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java (hunk 20 above)
Comment (OndrejKotek) [reply testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: Thank you. But it seem it was a wrong guess, this test case is still failing on some combinations.

Comment (OndrejKotek) [reply testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: IIUIC, `LdapRealmTestCase` and `FilesystemRealmTestCase` are also affected, but there are no related negative tests, hence the setup failure is not visible and tests pass with unsecured configuration.

Comment (OndrejKotek) [reply testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/FilesystemRealmEncryptedTestCase.java:103]: Maybe we can workaround this by renaming the `from-roles-attribute` resource, and create a new Jira issue for the other test cases.

**Comment by Ashpan** [comment]
Created PR to address clean tear down for `LdapRealmTestCase`
https://github.com/wildfly/wildfly/pull/15347

**Comment by bstansberry** [comment]
@Ashpan It seems the test fails when run against a bootable jar.
